blockchain/spam_throttler.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2021 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package blockchain
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/rcrowley/go-metrics"
+
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+)
+
+// TODO-Klaytn: move these variables into TxPool when BlockChain struct contains a TxPool interface
+// spamThrottler need to be accessed by both of TxPool and BlockChain.
+var (
+	spamThrottler   *throttler = nil
+	spamThrottlerMu            = new(sync.RWMutex)
+)
+
+var (
+	thresholdGauge           = metrics.NewRegisteredGauge("txpool/throttler/threshold", nil)
+	candidateSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/candidate/size", nil)
+	throttledSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/throttled/size", nil)
+	allowedSizeGauge         = metrics.NewRegisteredGauge("txpool/throttler/allowed/size", nil)
+	throttlerUpdateTimeGauge = metrics.NewRegisteredGauge("txpool/throttler/update/time", nil)
+	throttlerDropCount       = metrics.NewRegisteredCounter("txpool/throttler/dropped/count", nil)
+)
+
+type throttler struct {
+	config *ThrottlerConfig
+
+	candidates map[common.Address]int  // throttle candidates with spam weight. Not for concurrent use
+	throttled  map[common.Address]int  // throttled addresses with throttle time. Requires mu.lock for concurrent use
+	allowed    map[common.Address]bool // white listed addresses. Requires mu.lock for concurrent use
+	mu         *sync.RWMutex           // mutex for throttled and allowed
+
+	threshold  int
+	throttleCh chan *types.Transaction
+	quitCh     chan struct{}
+}
+
+type ThrottlerConfig struct {
+	ActivateTxPoolSize uint `json:"activate_tx_pool_size"`
+	TargetFailRatio    uint `json:"target_fail_ratio"`
```

Comment:
@aidan-kwon 
Very minor thing, but how about changing the word Ratio to Percentile? Maybe Percentile is more adequate word I think.
You can ignore this comment though :) 

If there is a percentage `50%`, it can be represented as
Ratio representation: `0.5`
Percentile representation: `50`

Reply:
Can I change this in another PR? Since this PR is included in v1.7.3 already. I want to change the name in the next version. The late change may not result in compatibility issues because it is a prototype feature only for CCOs. 
Thank you for the comment!

Reply:
@aidan-kwon 
Sure, why not?
Thanks aidan.